src/qibo/backends/_clifford_operations.py

```diff
@@ -0,0 +1,445 @@
+from functools import reduce
+
+import numpy as np
+
+
+def _get_rxz(symplectic_matrix, nqubits):
+    return (
+        symplectic_matrix[:, -1],
+        symplectic_matrix[:, :nqubits],
+        symplectic_matrix[:, nqubits:-1],
+    )
+
+
+def I(symplectic_matrix, q, nqubits):
+    return symplectic_matrix
+
+
+def H(symplectic_matrix, q, nqubits):
+    r, x, z = _get_rxz(symplectic_matrix, nqubits)
+    symplectic_matrix[:, -1] = r ^ (x[:, q] & z[:, q])
+    symplectic_matrix[:, [q, nqubits + q]] = symplectic_matrix[:, [nqubits + q, q]]
+    return symplectic_matrix
+
+
+def CNOT(symplectic_matrix, control_q, target_q, nqubits):
+    ind_zt = nqubits + target_q
+    ind_zc = nqubits + control_q
+    r = symplectic_matrix[:, -1]
+    xcq = symplectic_matrix[:, control_q]
+    xtq = symplectic_matrix[:, target_q]
+    ztq = symplectic_matrix[:, ind_zt]
+    zcq = symplectic_matrix[:, ind_zc]
+    symplectic_matrix[:, -1] = r ^ (xcq & ztq) & (xtq ^ ~zcq)
+    symplectic_matrix[:, target_q] = xtq ^ xcq
+    symplectic_matrix[:, ind_zc] = zcq ^ ztq
+    return symplectic_matrix
+
+
+def CZ(symplectic_matrix, control_q, target_q, nqubits):
+    """Decomposition --> H-CNOT-H"""
+    ind_zt = nqubits + target_q
+    ind_zc = nqubits + control_q
+    r = symplectic_matrix[:, -1]
+    xcq = symplectic_matrix[:, control_q]
+    xtq = symplectic_matrix[:, target_q]
+    ztq = symplectic_matrix[:, ind_zt]
+    zcq = symplectic_matrix[:, ind_zc]
+    ztq_xor_xcq = ztq ^ xcq
+    symplectic_matrix[:, -1] = (
+        r ^ (xtq & ztq) ^ (xcq & xtq & (ztq ^ ~zcq)) ^ (xtq & ztq_xor_xcq)
+    )
+    z_control_q = xtq ^ zcq
+    z_target_q = ztq_xor_xcq
+    symplectic_matrix[:, ind_zc] = z_control_q
+    symplectic_matrix[:, ind_zt] = z_target_q
+    return symplectic_matrix
+
+
+def S(symplectic_matrix, q, nqubits):
+    r, x, z = _get_rxz(symplectic_matrix, nqubits)
+    symplectic_matrix[:, -1] = r ^ (x[:, q] & z[:, q])
```

Comment:
Btw, in cases like this one, you already have `r` around. Consider that the `r` defined above is not a copy of a slice of the symplectic matrix, but a sliced view on it.
```suggestion
    r[:] = r ^ (x[:, q] & z[:, q])
```

(essentially, you could this for as long as you are using `__setitem__`, while it would not work if you're just redefining the variable, that would be instead overwritten)